pkg/storaged/resize.jsx

```diff
@@ -0,0 +1,441 @@
+            };
+            shrink_excuse = _("Stratis blockdevs can not be made smaller");
+        } else if (block.IdUsage == 'raid') {
+            info = { };
+            shrink_excuse = grow_excuse = _("Physical volumes can not be resized here.");
+        } else if (client.legacy_vdo_overlay.find_by_backing_block(block)) {
+            info = {
+                can_shrink: false,
+                can_grow: true,
+                grow_needs_unmount: false
+            };
+            shrink_excuse = _("VDO backing devices can not be made smaller");
+        } else {
+            info = {
+                can_shrink: false,
+                can_grow: true,
+                grow_needs_unmount: true
+            };
+            shrink_excuse = _("Unrecognized data can not be made smaller here.");
+        }
+        if (to_fit) {
+            // Shrink to fit doesn't need to resize the content
+            shrink_excuse = null;
+        }
+    } else {
+        info = { };
+        shrink_excuse = grow_excuse = _("This volume needs to be activated before it can be resized.");
+    }
+
+    return { info, shrink_excuse, grow_excuse };
+}
+
+export function free_space_after_part(client, part) {
+    const parts = get_partitions(client, client.blocks[part.Table]);
+
+    function find_it(parts) {
+        for (const p of parts) {
+            if (p.type == "free" && p.start == part.Offset + part.Size)
+                return p.size;
+            if (p.type == "container") {
+                const s = find_it(p.partitions);
+                if (s)
+                    return s;
+            }
+        }
+        return false;
+    }
+
+    return find_it(parts) || 0;
+}
+
+export function grow_dialog(client, lvol_or_part, info, to_fit) {
+    let title, block, name, orig_size, max_size, allow_infinite, round_size;
+
+    if (lvol_or_part.iface == "org.freedesktop.UDisks2.LogicalVolume") {
+        const vgroup = client.vgroups[lvol_or_part.VolumeGroup];
+        const pool = client.lvols[lvol_or_part.ThinPool];
+
+        title = _("Grow logical volume");
+        block = client.lvols_block[lvol_or_part.path];
+        name = lvol_or_part.Name;
+        orig_size = lvol_or_part.Size;
+        max_size = pool ? pool.Size * 3 : lvol_or_part.Size + vgroup.FreeSize;
+        allow_infinite = !!pool;
+        round_size = vgroup.ExtentSize;
+    } else {
+        title = _("Grow partition");
+        block = client.blocks[lvol_or_part.path];
+        name = block_name(block);
+        orig_size = lvol_or_part.Size;
+        max_size = lvol_or_part.Size + free_space_after_part(client, lvol_or_part);
+        allow_infinite = false;
+        round_size = 1024 * 1024;
+    }
+
+    const usage = get_active_usage(client, block && info.grow_needs_unmount ? block.path : null, _("grow"));
+
+    if (usage.Blocking) {
+        dialog_open({
+            Title: cockpit.format(_("$0 is in use"), name),
+            Body: BlockingMessage(usage)
```

Comment:
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19385-20230926-133853-59d8b5ea-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)